### PR TITLE
[debugger] Fix a step that becomes a go

### LINKED ISSUE #109785: Step becomes a "go" in .NET 9
### Description

.NET 9 contains a regression where ICorDebugStepper's are not always firing. The issue is non-deterministic, and causes intermittent failures in our debugging tests, and we have had a few customers reports of the problem.

### Reproduction Steps

A Microsoft employee can reproduce this problem by:

1. Enlisting in the 'Concord' repo
2. Following instructs in '<RepoRoot>/glass.md' for setting up a local test environment
3. Open Test Explorer
3. Enter “CSAsyncStepOverinTryBlock-NetCore-Latest-x64” in the filter box
4. Navigate to the test
5. Right click and invoke “Run Until Failure”

### Expected behavior

Stepper should complete

### Actual behavior

Stepper sometimes fails to detect when it is complete and this the debugger behaves as if there was no stepper

### Regression?

This regressed between .NET 8 and .NET 9

### Known Workarounds

None

### Configuration

So far all of the reports are from Windows, but I don't know if this is actually a causal connection.
We have at least mostly seen this on x64, but again, I am not sure if that is actually a causal connection or just because that is the most frequent configuration

### Other information

_No response_

### LINKED ISSUE #109812: `foreach` will jump to other code in `dotnet9`
### Description

When I upgrade the project to `Dotnet9` and `foreach` the list, I will jump to other code by myself after the list is not fully `foreach`.
dotnet8
https://github.com/user-attachments/assets/5bd3c4cb-2c7f-4b53-ae2f-adc318f6ee80

dotnet9
https://github.com/user-attachments/assets/9ddfc90d-0000-4bc7-a177-d962416d2d83

### Reproduction Steps

1. make a List<T>
2. add some thing
3. foreach it

### Expected behavior

foreach every item

### Actual behavior

foreach some item and exit all code

### Regression?

dotnet9

### Known Workarounds

_No response_

### Configuration

windows x64

### Other information

_No response_

### LINKED ISSUE #109885: Step Over sometimes executes a lot more than one statement
### Description

When I'm debugging code, I use F10 (Step Over) to walk through one line of code at a time.
In .NET 8x, this worked fine.
In .NET 9x, sometimes it seems like it runs a TON more code until it hits a breakpoint.

I suspect I might be running into some variant of https://github.com/dotnet/runtime/issues/109785 and I specifically notice this happening in ForEach code as described in https://github.com/dotnet/runtime/issues/109812 .

I see this is slated to be fixed in .NET 10.0, however, this is a major annoyance and I am very hopeful this will be approved for servicing.

### Reproduction Steps

Not sure.

### Expected behavior

"Step Over" works like .NET 8.0

### Actual behavior

"Step Over" sometimes works like "Continue"

### Regression?

Yes.

### Known Workarounds

Use .NET 8.0

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by thaystg**
After some tests we were able to understand that when we are stepping and the single step is enabled we can get on TriggerSingleStep in a Native code and then disable the single stepping wrongly.
This is the call stack that we see when the error happens:
image

The goal of this PR is to avoid disabling the single step when we get in this scenario.
So we added two protections on ThreadSuspend to avoid an APC when we are single stepping. And we also added a protection on controller.cpp to avoid disabling single step in this scenario.

Fixes https://github.com/dotnet/runtime/issues/109785
Fixes https://github.com/dotnet/runtime/issues/109812
Fixes https://github.com/dotnet/runtime/issues/109885

### CODE REVIEW

**Review by tommcdon — APPROVED** [review]

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Review by VSadov — APPROVED** [review]
LGTM, Thanks!

File: src/coreclr/vm/threadsuspend.cpp [hunk 1]
```diff
@@ -5733,7 +5733,9 @@ BOOL CheckActivationSafePoint(SIZE_T ip)
 
     // The criteria for safe activation is to be running managed code.
     // Also we are not interested in handling interruption if we are already in preemptive mode.
-    BOOL isActivationSafePoint = pThread != NULL &&
+    // Also we are not interested in handling interruption if we are single stepping
+    BOOL isActivationSafePoint = pThread != NULL && 
+        (pThread->m_StateNC & Thread::TSNC_DebuggerIsStepping) == 0 &&
```
Comment (mikelle-rogers) [inline src/coreclr/vm/threadsuspend.cpp:5738]: How come you used `pThread->m_stateNC` here, but in the change below, you just used `m_StateNC`?

File: src/coreclr/vm/threadsuspend.cpp [hunk 2]
```diff
@@ -5733,7 +5733,9 @@ BOOL CheckActivationSafePoint(SIZE_T ip)
 
     // The criteria for safe activation is to be running managed code.
     // Also we are not interested in handling interruption if we are already in preemptive mode.
-    BOOL isActivationSafePoint = pThread != NULL &&
+    // Also we are not interested in handling interruption if we are single stepping
```
Comment (mikelle-rogers) [inline src/coreclr/vm/threadsuspend.cpp:5736]: Nit: Maybe consider simplifying the three lines of comments to: 
// The criteria for safe activation is to be running managed code.
// Not interested in handling interruption if we are already in preemptive mode nor if we are single stepping

File: src/coreclr/debug/ee/controller.cpp [hunk 3]
```diff
@@ -7585,7 +7585,15 @@ bool DebuggerStepper::TriggerSingleStep(Thread *thread, const BYTE *ip)
     if (!g_pEEInterface->IsManagedNativeCode(ip))
     {
         LOG((LF_CORDB,LL_INFO10000, "DS::TSS: not in managed code, Returning false (case 0)!\n"));
-        DisableSingleStep();
+        // Sometimes we can get here with a callstack that is coming from an APC
+        // this will disable the single stepping and wrongly resume an app that the user
+        // is stepping.
```
Comment (mikelle-rogers) [inline src/coreclr/debug/ee/controller.cpp:7590]:
```suggestion
        // is stepping through.
```

File: src/coreclr/vm/threadsuspend.cpp (hunk 1 above)
Comment (thaystg) [reply src/coreclr/vm/threadsuspend.cpp:5737]: Because the other change is inside the class Thread in the method InjectActivation.

**Comment by thaystg** [comment]
The failure is not related with this PR: https://github.com/dotnet/runtime/issues/110285

**Pull request merged**